lib/tty/key.c

```diff
@@ -1718,6 +1746,117 @@ get_key_code (int no_delay)
     {
         gboolean bad_seq;
 
+        // May be a complex sequence
+    complex_seq:
+        while (pending_keys != NULL && pending_keys[0] == ESC_CHAR && pending_keys[1] == '[')
+        {
+            tty_nodelay (TRUE);
+            while ((c = tty_lowlevel_getch ()) >= 0)
+                push_char (c);
+            tty_nodelay (FALSE);
+
+            seq_complex = pending_keys;
+            pending_keys = seq_append = NULL;
+
+            // Check for complex sequence until any KEY_* constant (above 8bit)
+            for (i = 0; seq_complex[i] != '\0'; i++)
+            {
+                if (seq_complex[i] > 0xFF)
+                {
+                    pending_keys = seq_append = &seq_complex[i];
+                    break;
+                }
+                seq_char_buffer[i] = seq_complex[i];
```

Comment:
this is not robust against excessively long sequences. if we assume that the terminal may be malicious (ansi bomb or some such?), that would be a security hole.